File: testers/mariadb/scala/src/scala/testdb/DSLTest.scala
package testdb

import dev.typr.foundations.data.{Json, Uint1, Uint2, Uint4, Uint8}
import dev.typr.foundations.data.maria.{Inet4, Inet6}
import dev.typr.foundations.Bijection
import org.scalatest.funsuite.AnyFunSuite
import testdb.mariatest.*
import testdb.mariatest_identity.*

import java.math.BigInteger
import java.time.{LocalDate, LocalDateTime, LocalTime, Year}

class DSLTest extends AnyFunSuite {
  val mariatestRepo: MariatestRepoImpl = new MariatestRepoImpl
  val identityRepo: MariatestIdentityRepoImpl = new MariatestIdentityRepoImpl

  def createSampleRow(id: Int): MariatestRow = MariatestRow(
    tinyintCol = 127.toByte,
    smallintCol = 32767.toShort,
    mediumintCol = 8388607,
    intCol = MariatestId(id),
    bigintCol = 9223372036854775807L,
    tinyintUCol = Uint1.of(255),
    smallintUCol = Uint2.of(65535),
    mediumintUCol = Uint4.of(16777215L),
    intUCol = Uint4.of(4294967295L),
    bigintUCol = Uint8.of(new BigInteger("18446744073709551615")),
    decimalCol = BigDecimal("12345.67"),
    numericCol = BigDecimal("9876.5432"),
    floatCol = 3.14f,
    doubleCol = 2.718281828,
    boolCol = true,
    bitCol = Array(0xff.toByte),
    bit1Col = Array(0x01.toByte),
    charCol = "char_val  ",
    varcharCol = "varchar_value",
    tinytextCol = "tinytext",
    textCol = "text content",
    mediumtextCol = "mediumtext content",
    longtextCol = "longtext content",
    binaryCol = Array(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15),
    varbinaryCol = Array(1, 2, 3),
    tinyblobCol = Array(4, 5, 6),
    blobCol = Array(7, 8, 9),
    mediumblobCol = Array(10, 11, 12),
    longblobCol = Array(13, 14, 15),
    dateCol = LocalDate.of(2025, 1, 15),
    timeCol = LocalTime.of(14, 30, 45),
    timeFspCol = LocalTime.of(14, 30, 45, 123456000),
    datetimeCol = LocalDateTime.of(2025, 1, 15, 14, 30, 45),
    datetimeFspCol = LocalDateTime.of(2025, 1, 15, 14, 30, 45, 123456000),
    timestampCol = LocalDateTime.now(),
    timestampFspCol = LocalDateTime.now(),
    yearCol = Year.of(2025),
    setCol = XYZSet.fromString("x,y"),
    jsonCol = Json("{\"key\": \"value\"}"),
    inet4Col = new Inet4("192.168.1.1"),
    inet6Col = new Inet6("::1")
  )

  test("selectWithWhere") {
    withConnection {
      val row1 = mariatestRepo.insert(createSampleRow(1001).copy(varcharCol = "test_varchar_dsl"))

      val results = mariatestRepo.select
        .where(m => m.intCol.isEqual(row1.intCol))
        .where(m => m.varcharCol.isEqual("test_varchar_dsl"))
        .toList

      val _ = assert(results.size == 1)
      val _ = assert(results.head.intCol == row1.intCol)
      assert(results.head.varcharCol == "test_varchar_dsl")
    }
  }

  test("selectWithOrdering") {
    withConnection {
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("DSL_Zulu"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("DSL_Alpha"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("DSL_Mike"))

      val results = identityRepo.select
        .where(m => m.name.isEqual("DSL_Zulu").or(m.name.isEqual("DSL_Alpha")).or(m.name.isEqual("DSL_Mike")))
        .orderBy(m => m.name.asc)
        .toList

      val _ = assert(results.size >= 3)
      val sorted = results.filter(r => r.name.startsWith("DSL_"))
      assert(sorted.head.name == "DSL_Alpha")
    }
  }

  test("selectWithOrderByDesc") {
    withConnection {
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("DSLDescA"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("DSLDescB"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("DSLDescC"))

      val results = identityRepo.select
        .where(m => m.name.like("DSLDesc%", Bijection.asString()))
        .orderBy(m => m.name.desc)
        .toList

      val _ = assert(results.size == 3)
      val _ = assert(results.head.name == "DSLDescC")
      val _ = assert(results(1).name == "DSLDescB")
      assert(results(2).name == "DSLDescA")
    }
  }

  test("selectWithLimit") {
    withConnection {
      for (i <- 0 until 10) {
        val _ = identityRepo.insert(MariatestIdentityRowUnsaved(s"Limit$i"))
      }

      val results = identityRepo.select
        .where(m => m.name.like("Limit%", Bijection.asString()))
        .limit(3)
        .toList

      assert(results.size == 3)
    }
  }

  test("selectWithOffset") {
    withConnection {
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("OffsetA"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("OffsetB"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("OffsetC"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("OffsetD"))

      val results = identityRepo.select
        .where(m => m.name.like("Offset%", Bijection.asString()))
        .orderBy(m => m.name.asc)
        .offset(2)
        .limit(10)
        .toList

      val _ = assert(results.size == 2)
      val _ = assert(results.head.name == "OffsetC")
      assert(results(1).name == "OffsetD")
    }
  }

  test("selectWithCount") {
    withConnection {
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("CountA"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("CountB"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("CountC"))

      val count = identityRepo.select
        .where(m => m.name.like("Count%", Bijection.asString()))
        .count

      assert(count == 3)
    }
  }

  test("selectWithGreaterThan") {
    withConnection {
      val row1 = identityRepo.insert(MariatestIdentityRowUnsaved("GT1"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("GT2"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("GT3"))

      val results = identityRepo.select
        .where(m => m.id.greaterThan(row1.id))
        .where(m => m.name.like("GT%", Bijection.asString()))
        .toList

      assert(results.size == 2)
    }
  }

  test("selectWithLike") {
    withConnection {
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("LikeTest_ABC"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("LikeTest_XYZ"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("OtherName"))

      val results = identityRepo.select
        .where(m => m.name.like("LikeTest%", Bijection.asString()))
        .toList

      assert(results.size == 2)
    }
  }

  test("selectWithIn") {
    withConnection {
      val row1 = identityRepo.insert(MariatestIdentityRowUnsaved("InTest1"))
      val _ = identityRepo.insert(MariatestIdentityRowUnsaved("InTest2"))
      val row3 = identityRepo.insert(MariatestIdentityRowUnsaved("InTest3"))

      val results = identityRepo.select
        .where(m => m.id.in(row1.id, row3.id))
        .toList

      assert(results.size == 2)
    }
  }

  test("selectWithProjection") {
    withConnection {
      val row = identityRepo.insert(MariatestIdentityRowUnsaved("ProjectionTest"))

      val results = identityRepo.select
        .where(m => m.id.isEqual(row.id))
        .map(m => m.name.tupleWith(m.id))
        .toList

      val _ = assert(results.size == 1)
      val _ = assert(results.head._1() == "ProjectionTest")
      assert(results.head._2() == row.id)
    }
  }

  test("deleteWithDSL") {
    withConnection {
      val row = identityRepo.insert(MariatestIdentityRowUnsaved("ToDeleteDSL"))

      val beforeCount = identityRepo.select
        .where(m => m.id.isEqual(row.id))
        .count
      val _ = assert(beforeCount == 1)

      val _ = identityRepo.deleteById(row.id)

      val afterCount = identityRepo.select
        .where(m => m.id.isEqual(row.id))
        .count
      assert(afterCount == 0)
    }
  }
}

File: testers/mariadb/java/src/java/testdb/DSLTest.java
package testdb;

import static org.junit.Assert.*;

import dev.typr.foundations.Bijection;
import dev.typr.foundations.data.maria.Inet4;
import dev.typr.foundations.data.maria.Inet6;
import java.math.BigDecimal;
import java.time.Year;
import java.util.Random;
import org.junit.Test;
import testdb.mariatest.*;
import testdb.mariatest_identity.*;

/** Tests for DSL (Domain-Specific Language) operations in MariaDB. */
public class DSLTest {
  private final MariatestRepoImpl mariatestRepo = new MariatestRepoImpl();
  private final MariatestIdentityRepoImpl identityRepo = new MariatestIdentityRepoImpl();
  private final TestInsert testInsert = new TestInsert(new Random(141839624));

  @Test
  public void testSelectWithWhere() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row1 =
              testInsert
                  .Mariatest(
                      new byte[] {0x01},
                      new byte[] {0x01},
                      new byte[16],
                      new byte[] {1, 2},
                      new byte[] {3, 4},
                      new byte[] {5, 6},
                      new byte[] {7, 8},
                      new byte[] {9, 10},
                      Year.of(2025),
                      new Inet4("192.168.1.1"),
                      new Inet6("::1"))
                  .with(r -> r.withVarcharCol("test_varchar"))
                  .insert(c);

          // Use DSL to select
          var query =
              mariatestRepo
                  .select()
                  .where(m -> m.intCol().isEqual(row1.intCol()))
                  .where(m -> m.varcharCol().isEqual("test_varchar"));

          var results = query.toList(c);
          assertEquals(1, results.size());
          assertEquals(row1.intCol(), results.get(0).intCol());
          assertEquals("test_varchar", results.get(0).varcharCol());
        });
  }

  @Test
  public void testSelectWithOrdering() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          identityRepo.insert(new MariatestIdentityRowUnsaved("Zulu"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("Alpha"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("Mike"), c);

          // Select with ordering
          var query =
              identityRepo
                  .select()
                  .where(
                      m ->
                          m.name()
                              .isEqual("Zulu")
                              .or(
                                  m.name().isEqual("Alpha"),
                                  dev.typr.foundations.Bijection.asBool())
                              .or(
                                  m.name().isEqual("Mike"),
                                  dev.typr.foundations.Bijection.asBool()))
                  .orderBy(m -> m.name().asc());

          var results = query.toList(c);
          assertTrue(results.size() >= 3);
          // First result should be "Alpha" (alphabetically first)
          var sorted = results.stream().filter(r -> r.name().equals("Alpha")).findFirst();
          assertTrue(sorted.isPresent());
        });
  }

  @Test
  public void testSelectWithOrderByDesc() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          identityRepo.insert(new MariatestIdentityRowUnsaved("OrderDescA"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("OrderDescB"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("OrderDescC"), c);

          // Select with descending order
          var query =
              identityRepo
                  .select()
                  .where(m -> m.name().like("OrderDesc%", Bijection.asString()))
                  .orderBy(m -> m.name().desc());

          var results = query.toList(c);
          assertEquals(3, results.size());
          assertEquals("OrderDescC", results.get(0).name());
          assertEquals("OrderDescB", results.get(1).name());
          assertEquals("OrderDescA", results.get(2).name());
        });
  }

  @Test
  public void testSelectWithLimit() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          for (int i = 0; i < 10; i++) {
            identityRepo.insert(new MariatestIdentityRowUnsaved("Limit" + i), c);
          }

          // Select with limit
          var query =
              identityRepo
                  .select()
                  .where(m -> m.name().like("Limit%", Bijection.asString()))
                  .limit(3);

          var results = query.toList(c);
          assertEquals(3, results.size());
        });
  }

  @Test
  public void testSelectWithOffset() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          identityRepo.insert(new MariatestIdentityRowUnsaved("OffsetA"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("OffsetB"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("OffsetC"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("OffsetD"), c);

          // Select with offset
          var query =
              identityRepo
                  .select()
                  .where(m -> m.name().like("Offset%", Bijection.asString()))
                  .orderBy(m -> m.name().asc())
                  .offset(2)
                  .limit(10);

          var results = query.toList(c);
          assertEquals(2, results.size());
          assertEquals("OffsetC", results.get(0).name());
          assertEquals("OffsetD", results.get(1).name());
        });
  }

  @Test
  public void testSelectWithCount() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          identityRepo.insert(new MariatestIdentityRowUnsaved("CountA"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("CountB"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("CountC"), c);

          // Count
          var query =
              identityRepo.select().where(m -> m.name().like("Count%", Bijection.asString()));

          long count = query.count(c);
          assertEquals(3, count);
        });
  }

  @Test
  public void testSelectWithGreaterThan() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row1 = identityRepo.insert(new MariatestIdentityRowUnsaved("GT1"), c);
          var row2 = identityRepo.insert(new MariatestIdentityRowUnsaved("GT2"), c);
          var row3 = identityRepo.insert(new MariatestIdentityRowUnsaved("GT3"), c);

          // Select with greater than
          var query =
              identityRepo
                  .select()
                  .where(m -> m.id().greaterThan(row1.id()))
                  .where(m -> m.name().like("GT%", Bijection.asString()));

          var results = query.toList(c);
          assertEquals(2, results.size());
        });
  }

  @Test
  public void testSelectWithLessThan() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row1 = identityRepo.insert(new MariatestIdentityRowUnsaved("LT1"), c);
          var row2 = identityRepo.insert(new MariatestIdentityRowUnsaved("LT2"), c);
          var row3 = identityRepo.insert(new MariatestIdentityRowUnsaved("LT3"), c);

          // Select with less than
          var query =
              identityRepo
                  .select()
                  .where(m -> m.id().lessThan(row3.id()))
                  .where(m -> m.name().like("LT%", Bijection.asString()));

          var results = query.toList(c);
          assertEquals(2, results.size());
        });
  }

  @Test
  public void testSelectWithLike() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          identityRepo.insert(new MariatestIdentityRowUnsaved("LikeTest_ABC"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("LikeTest_XYZ"), c);
          identityRepo.insert(new MariatestIdentityRowUnsaved("OtherName"), c);

          // Select with LIKE
          var query =
              identityRepo.select().where(m -> m.name().like("LikeTest%", Bijection.asString()));

          var results = query.toList(c);
          assertEquals(2, results.size());
        });
  }

  @Test
  public void testSelectWithIn() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row1 = identityRepo.insert(new MariatestIdentityRowUnsaved("InTest1"), c);
          var row2 = identityRepo.insert(new MariatestIdentityRowUnsaved("InTest2"), c);
          var row3 = identityRepo.insert(new MariatestIdentityRowUnsaved("InTest3"), c);

          // Select with IN
          var query = identityRepo.select().where(m -> m.id().in(row1.id(), row3.id()));

          var results = query.toList(c);
          assertEquals(2, results.size());
        });
  }

  @Test
  public void testSelectWithProjection() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row = identityRepo.insert(new MariatestIdentityRowUnsaved("ProjectionTest"), c);

          // Select with projection (map to specific columns)
          var query =
              identityRepo
                  .select()
                  .where(m -> m.id().isEqual(row.id()))
                  .map(m -> m.name().tupleWith(m.id()));

          var results = query.toList(c);
          assertEquals(1, results.size());
          assertEquals("ProjectionTest", results.get(0)._1());
          assertEquals(row.id(), results.get(0)._2());
        });
  }

  @Test
  public void testSelectWithMultipleConditions() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row =
              testInsert
                  .Mariatest(
                      new byte[] {0x01},
                      new byte[] {0x01},
                      new byte[16],
                      new byte[] {1, 2},
                      new byte[] {3, 4},
                      new byte[] {5, 6},
                      new byte[] {7, 8},
                      new byte[] {9, 10},
                      Year.of(2025),
                      new Inet4("10.0.0.1"),
                      new Inet6("fe80::1"))
                  .with(
                      r ->
                          r.withVarcharCol("multi_cond")
                              .withBoolCol(true)
                              .withDecimalCol(new BigDecimal("100.00")))
                  .insert(c);

          // Select with multiple conditions
          var query =
              mariatestRepo
                  .select()
                  .where(m -> m.varcharCol().isEqual("multi_cond"))
                  .where(m -> m.boolCol().isEqual(true));

          var results = query.toList(c);
          assertEquals(1, results.size());
        });
  }

  @Test
  public void testUpdateWithDSL() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row = identityRepo.insert(new MariatestIdentityRowUnsaved("ToUpdate"), c);

          // Update via repo (DSL update setValue pattern)
          var updated = row.withName("UpdatedName");
          identityRepo.update(updated, c);

          // Verify update
          var query = identityRepo.select().where(m -> m.id().isEqual(row.id()));

          var results = query.toList(c);
          assertEquals(1, results.size());
          assertEquals("UpdatedName", results.get(0).name());
        });
  }

  @Test
  public void testDeleteWithDSL() {
    MariaDbTestHelper.run(
        c -> {
          // Insert test data
          var row = identityRepo.insert(new MariatestIdentityRowUnsaved("ToDelete"), c);

          // Verify exists
          var beforeCount = identityRepo.select().where(m -> m.id().isEqual(row.id())).count(c);
          assertEquals(1, beforeCount);

          // Delete
          identityRepo.deleteById(row.id(), c);

          // Verify deleted
          var afterCount = identityRepo.select().where(m -> m.id().isEqual(row.id())).count(c);
          assertEquals(0, afterCount);
        });
  }
}
